### PR TITLE
Fix Issue 6106 - Keep track of changes during replace function

### PR DESCRIPTION
**Pull request by RazvanN7**
Although I basically duplicated the code in the 2 methods, I think this is preferable to computing the replacements when that is not needed.

### CODE REVIEW

File: std/array.d [hunk 1]
```diff
@@ -2016,11 +2019,40 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+        && (hasLength!R2 || isSomeString!R2))
+{
+    import std.algorithm.searching : find;
+
+    countChanges = 0;
+    if (from.empty) return subject;
+
+    auto balance = find(subject, from.save);
+    if (balance.empty)
+        return subject;
```
Comment (wilzbach) [inline std/array.d]: Not covered, Please install the CodeCov browser extension: https://github.com/codecov/browser-extension

File: std/array.d [hunk 2]
```diff
@@ -2016,11 +2019,40 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+
+    auto app = appender!(E[])();
```
Comment (wilzbach) [inline std/array.d]: `(ElementType!Subject)[]`

File: std/array.d [hunk 3]
```diff
@@ -2016,11 +2019,40 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
 ///
 @safe unittest
 {
     assert("Hello Wörld".replace("o Wö", "o Wo") == "Hello World");
     assert("Hello Wörld".replace("l", "h") == "Hehho Wörhd");
+
+    size_t countChanges;
+    assert("Sooo many oooos".replace("o", "ö", countChanges) == "Sööö many öööös");
+    assert(countChanges == 7); // o was replaced 7 times with ö
 }
```
Comment (wilzbach) [inline std/array.d]: 1) Test with `const string` as input
2) Test with a forward range:

```
assert("Hello Wörld".filter(_ => true).replace("o Wö", "o Wo") == "Hello World");
```

File: std/array.d [hunk 4]
```diff
@@ -2016,11 +2019,40 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
     return app.data;
 }
 
+/// ditto
+E[] replace(E, R1, R2)(E[] subject, R1 from, R2 to, ref size_t countChanges)
+if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
+        && (hasLength!R2 || isSomeString!R2))
```
Comment (wilzbach) [inline std/array.d]: Also `subject` should be `const` or `in`

Comment (wilzbach) [inline std/array.d]: (and of course this applies to the existing `replace` function as well).

File: std/array.d [hunk 5]
```diff
@@ -2016,11 +2019,40 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+    if (from.empty) return subject;
+
+    auto balance = find(subject, from.save);
+    if (balance.empty)
+        return subject;
+
+    auto app = appender!(E[])();
+    app.put(subject[0 .. subject.length - balance.length]);
+    app.put(to.save);
+    replaceInto(app, balance[from.length .. $], from, to, countChanges);
```
Comment (wilzbach) [inline std/array.d]: `balance.drop(from.length)`

**Comment by wilzbach** [comment]
Btw why doesn't Phobos have a lazy replace like [`substitute` in phobos-next](https://github.com/nordlow/phobos-next/blob/1992896e945111aa3f47eb94a42dcc57be4cc61b/src/substitution.d)?
The version in phobos-next is extremely cool, because if the replacement words are known at CT it's faster :)

**Comment by wilzbach** [comment]
> bt why doesn't Phobos have a lazy replace 

-> https://github.com/dlang/phobos/pull/5596

**Comment by RazvanN7** [comment]
@wilzbach After applying the modifications, when I try to compile I get the following error : 

std/uni.d(1611): Error: function std.array.replace!(string, string, string).replace is not nothrow
std/uni.d(1592): Error: nothrow function std.uni.genUnrolledSwitchSearch may throw
std/uni.d(1661): Error: CTFE failed because of previous errors in genUnrolledSwitchSearch
std/uni.d(1676): Error: template instance std.uni.switchUniformLowerBound!(binaryFun, const(uint)[], uint) error instantiating
std/uni.d(2197):        instantiated from here: sharMethod!("a <= b", const(uint)[], uint)
std/uni.d(1977):        instantiated from here: InversionList!(GcPolicy)

It seems that the function isn't nothrow anymore. I'm not sure how to solve this. I'll push what I've done and maybe you have an idea on what's going on

**Comment by wilzbach** [comment]
> It seems that the function isn't nothrow anymore. I'm not sure how to solve this. I'll push what I've done and maybe you have an idea on what's going on

So the following works on my machine. As you can see I did a few things:
- I removed the duplicated methods (I don't think the overhead of counting is measurable if there even exists one with a decent optimizer, feel free to measure )
- One needs to be careful with counting and `UTF` -> we could either do an optional `byUTF` iteration on `isNarrowString` or we simply defer this to existing logic in `findSplit`

However to play the devil's advocate here: `replace` is in `std.array` and thus its expected to work with arrays. OTOH making it generic is simply replacing `find` with `findSplit` and deferring the hard-coded slicing to `findSplit`

**Comment by RazvanN7** [comment]
@wilzbach Nice. Thanks. Looks good.

File: std/array.d [hunk 6]
```diff
@@ -1980,38 +1980,62 @@ if (isInputRange!RoR &&
...
+    {
+        auto res = findSplit(subject, from);
+    }
+
+    if (res[1].empty)
+        return subject;
+
+    auto app = appender!((ElementEncodingType!E)[])();
+    put(app, res[0]);
+    put(app, to.save);
```
Comment (wilzbach) [inline std/array.d]: Btw @RazvanN7 all the `.save` calls that you have inserted are only necessary if by-reference Ranges are supported. Currently there are no tests for this, hence maybe adding a test for it wouldn't be a bad idea ;-)
Tip: [`refRange`](https://dlang.org/phobos/std_range.html#.refRange) is your friend ;-)

**Comment by wilzbach** [comment]
> @wilzbach Nice. Thanks. Looks good.

As I modified quite a bit of this PR, I think it's better if I defer the review to someone else.

File: std/array.d [hunk 7]
```diff
@@ -2021,32 +2045,56 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
+
+/// ditto
+void replaceInto(E, Sink, R1, R2)(Sink sink, E subject, R1 from, R2 to)
+{
+    size_t countChanges = 0;
+    replaceInto(sink, subject, from, to, countChanges);
 }
 
 /// ditto
-void replaceInto(E, Sink, R1, R2)(Sink sink, E[] subject, R1 from, R2 to)
```
Comment (JackStouffer) [inline std/array.d]: Geez, when was this signature made? The front is the worst place to put the sink argument.

File: std/array.d [hunk 8]
```diff
@@ -2021,32 +2045,56 @@ if (isDynamicArray!(E[]) && isForwardRange!R1 && isForwardRange!R2
...
     assert("Hello Wörld".replace("o Wö", "o Wo") == "Hello World");
     assert("Hello Wörld".replace("l", "h") == "Hehho Wörhd");
+
+    size_t countChanges;
+    assert("Sooo many oooos".replace("o", "ö", countChanges) == "Sööö many öööös");
+    assert(countChanges == 7); // o was replaced 7 times with ö
+}
+
+/// ditto
+void replaceInto(E, Sink, R1, R2)(Sink sink, E subject, R1 from, R2 to)
```
Comment (JackStouffer) [inline std/array.d]: The least complicated signature should be first in the docs

**Pull request closed without merging**